engine/controller/controller_test.go

```diff
@@ -84,10 +87,167 @@ func TestNewAsyncControllerWithMap(t *testing.T) {
 }
 
 func TestSubmitRequest(t *testing.T) {
-	// TODO: actually implement a test
-	ctrl := NewAsyncController(1)
-	output, err := ctrl.SubmitRequest(&runtime.RunProps{})
-	if output != nil && err != nil {
-		t.Error()
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	happyCaseOutput := &runtime.RunOutput{
+		Stdout: "hello",
+		Stderr: "world",
+	}
+
+	tests := []struct {
+		name     string
+		runProps *runtime.RunProps
+		mockInfo []mockProps
+		want     ControllerRunOutput
+	}{
+		{
```

Comment:
These test cases can be improved to maybe generate a really, *really* long list of runners then execute on that.

There also should be integration tests making sure the controller works when the runner isn't actually a mock. That can be done in #55 though.